pilot/cmd/pilot-agent/status/server.go

```diff
@@ -149,67 +96,32 @@ func (s *Server) Run(ctx context.Context) {
 	// Wait for the agent to be shut down.
 	<-ctx.Done()
 	log.Info("Status server has successfully terminated")
+	s.notifyReady()
 }
 
-func (s *Server) handleReadyProbe(w http.ResponseWriter, _ *http.Request) {
-	err := s.ready.Check()
-
+// WaitForReady waits until the server is up and running and capable of receiving requests.
+func (s *Server) WaitForReady() {
 	s.mutex.Lock()
-	if err != nil {
-		w.WriteHeader(http.StatusServiceUnavailable)
+	defer s.mutex.Unlock()
 
-		log.Infof("Envoy proxy is NOT ready: %s", err.Error())
-		s.lastProbeSuccessful = false
-	} else {
-		w.WriteHeader(http.StatusOK)
-
-		if !s.lastProbeSuccessful {
-			log.Info("Envoy proxy is ready")
-		}
-		s.lastProbeSuccessful = true
+	if !s.ready {
+		s.cond.Wait()
 	}
-	s.mutex.Unlock()
 }
 
-func (s *Server) handleAppProbe(w http.ResponseWriter, req *http.Request) {
-	// Validate the request first.
-	path := req.URL.Path
-	if !strings.HasPrefix(path, "/") {
-		path = "/" + req.URL.Path
-	}
-	prober, exists := s.appKubeProbers[path]
-	if !exists {
-		log.Errorf("Prober does not exists url %v", path)
-		w.WriteHeader(http.StatusBadRequest)
-		w.Write([]byte(fmt.Sprintf("app prober config does not exists for %v", path)))
-		return
-	}
+// GetConfig provides access to the configuration for this server.
+func (s *Server) GetConfig() Config {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.config
+}
 
-	// Construct a request sent to the application.
-	httpClient := &http.Client{
-		// TODO: figure out the appropriate timeout?
-		Timeout: 10 * time.Second,
-	}
-	url := fmt.Sprintf("http://127.0.0.1:%v%s", prober.Port.IntValue(), prober.Path)
-	appReq, err := http.NewRequest("GET", url, nil)
-	if err != nil {
-		log.Errorf("Failed to create request to probe app %v, original url %v", err, path)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
-	}
-	for _, header := range prober.HTTPHeaders {
-		appReq.Header[header.Name] = []string{header.Value}
-	}
+func (s *Server) notifyReady() {
```

Comment:
Could you write some comments explaining this sync model? I'm not clear on which goroutines are waiting on this.

Reply:
done.